include/libtorrent/aux_/win_crypto_provider.hpp

```diff
@@ -0,0 +1,162 @@
+/*
+
+Copyright (c) 2016, Arvid Norberg
+All rights reserved.
+
+Redistribution and use in source and binary forms, with or without
+modification, are permitted provided that the following conditions
+are met:
+
+		* Redistributions of source code must retain the above copyright
+			notice, this list of conditions and the following disclaimer.
+		* Redistributions in binary form must reproduce the above copyright
+			notice, this list of conditions and the following disclaimer in
+			the documentation and/or other materials provided with the distribution.
+		* Neither the name of the author nor the names of its
+			contributors may be used to endorse or promote products derived
+			from this software without specific prior written permission.
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
+LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+POSSIBILITY OF SUCH DAMAGE.
+
+*/
+
+#ifndef TORRENT_WIN_CRYPTO_PROVIDER_HPP
+#define TORRENT_WIN_CRYPTO_PROVIDER_HPP
+
+#include "libtorrent/config.hpp"
+#include "libtorrent/error_code.hpp"
+
+#include <windows.h>
+#include <wincrypt.h>
+
+namespace libtorrent { namespace aux
+{
+	template <DWORD ProviderType>
+	struct crypt_provider
+	{
+	protected:
+		static HCRYPTPROV acquire_provider(DWORD provider_type)
+		{
+			HCRYPTPROV ret;
+			if (CryptAcquireContext(&ret, nullptr, nullptr, provider_type
+				, CRYPT_VERIFYCONTEXT) == false)
+			{
+#ifndef BOOST_NO_EXCEPTIONS
+				throw system_error(error_code(GetLastError(), system_category()));
+#else
+				std::terminate();
+#endif
+			}
+			return ret;
+		}
+		static HCRYPTPROV get_provider()
+		{
+			static HCRYPTPROV provider = acquire_provider(ProviderType);
+			return provider;
+		}
+	};
+
+	struct crypt_random: crypt_provider<PROV_RSA_FULL>
+	{
+		static void generate(span<char> buffer)
+		{
+			if (!CryptGenRandom(get_provider(), int(buffer.size())
+				, reinterpret_cast<BYTE*>(buffer.data())))
+			{
+#ifndef BOOST_NO_EXCEPTIONS
+				throw system_error(error_code(GetLastError(), system_category()));
+#else
+				std::terminate();
+#endif
+			}
+		}
+	};
+
+	template <ALG_ID AlgId, DWORD ProviderType>
```

Comment:
``AlgId`` doesn't need to be a template parameter. just pass it in to the constructor.